type-metadata/src/lib.rs

```diff
@@ -1,7 +1,346 @@
+	Tuple(TupleIdent),
+	Option(OptionIdent),
+	Result(ResultIdent),
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct Namespace(Vec<&'static str>);
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct CustomIdent {
+	pub name: &'static str,
+	pub namespace: Namespace,
+	pub type_params: Vec<IdentKind>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ArrayIdent {
+	pub len: u16,
+	pub type_param: Box<IdentKind>,
+}
+impl ArrayIdent {
+	fn new(len: u16, type_param: IdentKind) -> Self {
+		ArrayIdent {
+			len,
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct TupleIdent {
+	pub type_params: Option<Vec<IdentKind>>,
+}
+impl TupleIdent {
+	fn new(type_params: Option<Vec<IdentKind>>) -> Self {
+		TupleIdent { type_params }
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct SliceIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl SliceIdent {
+	fn new(type_param: IdentKind) -> Self {
+		SliceIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct OptionIdent {
+	pub type_param: Box<IdentKind>,
+}
+impl OptionIdent {
+	fn new(type_param: IdentKind) -> Self {
+		OptionIdent {
+			type_param: Box::new(type_param),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug)]
+pub struct ResultIdent {
+	pub type_params: (Box<IdentKind>, Box<IdentKind>),
+}
+impl ResultIdent {
+	fn new(type_params: (IdentKind, IdentKind)) -> Self {
+		ResultIdent {
+			type_params: (Box::new(type_params.0), Box::new(type_params.1)),
+		}
+	}
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum TypeDef {
+	Primitive,
+	Struct(StructDef),
+	TupleStruct(TupleStructDef),
+	Enum(EnumDef),
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct Field {
+	pub name: &'static str,
+	pub ident: IdentKind,
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct StructDef(Vec<Field>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct TupleStructDef(Vec<IdentKind>);
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct DataVariant<T> {
+	pub name: &'static str,
+	pub index: u16,
+	pub struct_def: T,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub struct NoDataVariant {
+	pub name: &'static str,
+	pub index: u16,
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum VariantKind {
+	NoData(NoDataVariant),
+	Struct(DataVariant<StructDef>),
+	TupleStruct(DataVariant<TupleStructDef>),
+}
+#[derive(PartialEq, Eq, Debug)]
+pub struct EnumDef(Vec<VariantKind>);
+
+pub trait Metadata {
+	fn type_ident() -> IdentKind;
+
+	/// If the current type contains any other types, `type_def` would register their metadata into the given
+	/// `registry`. For instance, `<Option<MyStruct>>::type_def()` would register `MyStruct` metadata. All
+	/// implementation must register these contained types' metadata.
+	fn type_def(registry: &mut Registry) -> TypeDef;
+
+	fn register(registry: &mut Registry) {
```

Comment:
Same here: If you do not want to allow for a special treatment and this definition is good for all implementors then move this trait fn out into a free function that is generic over `T: Metadata` to make the trait simpler.

Reply:
Relocated `register` fn to `Registry`: https://github.com/cennznet/type-metadata/blob/716fecaf094ca5557c9ecc86d549485a5a19c266/type-metadata/src/registry.rs#L54-L56